File: ios/Classes/coins/ETH.swift
/*
 ETH
 */
import WalletCore

class ETH: Coin  {
    init() {
        super.init(name: "ETH", coinType: .ethereum)
    }
    override func signTransaction(path: String, txData: [String : Any], mnemonic: String, passphrase: String) -> String? {
        let privateKey = HDWallet(mnemonic: mnemonic, passphrase: passphrase)?.getKey(coin: self.coinType, derivationPath: path)
        let cmd = txData["cmd"] as? String
        
        if privateKey != nil {
            var input = EthereumSigningInput.with {
                $0.chainID = Data(hexString: txData["chainId"] as! String)!
                $0.nonce = Data(hexString: (txData["nonce"] as! String))!
                $0.gasPrice = Data(hexString: (txData["gasPrice"] as! String))!
                $0.gasLimit = Data(hexString: (txData["gasLimit"] as! String))!
                $0.toAddress = txData["toAddress"] as! String
                $0.privateKey = privateKey!.data
            }
            switch cmd {
                case "ERC20":
                    input.toAddress = txData["contractAddress"] as! String
                    input.transaction = EthereumTransaction.with {
                        $0.erc20Transfer = EthereumTransaction.ERC20Transfer.with {
                            $0.to = txData["toAddress"] as! String
                            $0.amount = Data(hexString: (txData["amount"] as! String))!
                        }
                    }
                    
                default:
                    input.toAddress = txData["toAddress"] as! String
                    input.transaction = EthereumTransaction.with {
                        $0.transfer = EthereumTransaction.Transfer.with {
                            $0.amount = Data(hexString: (txData["amount"] as! String))!
                        }
                    }
                
            }
            let sign: EthereumSigningOutput = AnySigner.sign(input: input, coin: coinType)
            return sign.encoded.hexString
        }
        return nil
    }
}

File: android/src/main/kotlin/africa/ejara/trustdart/coins/ETH.kt
import africa.ejara.trustdart.Coin
import com.google.protobuf.ByteString
import wallet.core.java.AnySigner
import africa.ejara.trustdart.Numeric
import africa.ejara.trustdart.utils.toHexByteArray
import wallet.core.jni.CoinType
import wallet.core.jni.HDWallet
import wallet.core.jni.proto.Ethereum.SigningOutput
import wallet.core.jni.proto.Ethereum

open class ETH: Coin("ETH", CoinType.ETHEREUM)  {

    override fun signTransaction(
        path: String,
        txData: Map<String, Any>,
        mnemonic: String,
        passphrase: String
    ): String? {
        val wallet = HDWallet(mnemonic, passphrase)
        val cmd: String? = txData["cmd"] as? String
        val signingInput = Ethereum.SigningInput.newBuilder()

        signingInput.apply {
            privateKey = ByteString.copyFrom(wallet.getKey(coinType, path).data())
            chainId = ByteString.copyFrom((txData["chainID"] as String).toHexByteArray())
            nonce = ByteString.copyFrom((txData["nonce"] as String).toHexByteArray())
            gasPrice = ByteString.copyFrom((txData["gasPrice"] as String).toHexByteArray())
            gasLimit = ByteString.copyFrom((txData["gasLimit"] as String).toHexByteArray())
        }
        when (cmd) {
            "ERC20" -> {

                var transaction = Ethereum.Transaction.newBuilder().apply {
                    erc20Transfer = Ethereum.Transaction.ERC20Transfer.newBuilder().apply {
                        to = txData["toAddress"] as String
                        amount = ByteString.copyFrom((txData["amount"] as String).toHexByteArray())
                    }.build()
                }.build()
                signingInput.setToAddress(txData["contractAddress"] as String)
                signingInput.setTransaction(transaction)
                
            }
            else -> {
                var transaction = Ethereum.Transaction.newBuilder().apply {
                    transfer = Ethereum.Transaction.Transfer.newBuilder().apply {
                        amount = ByteString.copyFrom((txData["amount"] as String).toHexByteArray())
                    }.build()
                }.build()

                signingInput.setToAddress(txData["toAddress"] as String)
                signingInput.setTransaction(transaction)
            }
        }
        val sign = AnySigner.sign(signingInput.build(), coinType, SigningOutput.parser())
        return Numeric.toHexString(sign.encoded.toByteArray())

    }

}